src/main/java/com/aws/greengrass/ipc/SubscribeToCertificateUpdatesOperationHandler.java

```diff
@@ -0,0 +1,190 @@
+        });
+    }
+
+
+    private CertificateType getCertificateTypeFromCertificateOptions(CertificateOptions certificateOptions) {
+        validateCertificateOptions(certificateOptions);
+        CertificateType certificateType = certificateOptions.getCertificateType();
+        if (certificateType == null) {
+            logger.atError().kv(COMPONENT_NAME, serviceName).log(NO_CERT_TYPE_ERROR);
+            throw new InvalidArgumentsError(NO_CERT_TYPE_ERROR);
+        }
+        return certificateType;
+    }
+
+    private void validateCertificateOptions(CertificateOptions certificateOptions) {
+        if (certificateOptions == null) {
+            logger.atError().kv(COMPONENT_NAME, serviceName).log(NO_CERT_OPTIONS_ERROR);
+            throw new InvalidArgumentsError(NO_CERT_OPTIONS_ERROR);
+        }
+    }
+
+    @SuppressWarnings("PMD.PreserveStackTrace")
+    private void forwardServerCertificates(X509Certificate certificate) {
+        CertificateUpdate certificateUpdate = new CertificateUpdate();
+        try {
+            certificateUpdate
+                    .withCertificate(EncryptionUtils.encodeToPem(PEM_BOUNDARY_CERTIFICATE, certificate.getEncoded()))
+                    .withCaCertificates(this.certificateManager.getCACertificates())
+                    .withPublicKey(
+                            EncryptionUtils.encodeToPem(PEM_BOUNDARY_PUBLIC_KEY, keyPair.getPublic().getEncoded()))
+                    .withPrivateKey(
+                            EncryptionUtils.encodeToPem(PEM_BOUNDARY_PRIVATE_KEY, keyPair.getPrivate().getEncoded()));
+        } catch (CertificateEncodingException | IOException | KeyStoreException e) {
+            logger.atError().cause(e).log("Unable to attach certificates to the response");
+            throw new ServiceError("Subscribe to certificate update failed. Check Greengrass log for details.");
+        }
+        CertificateUpdateEvent event = new CertificateUpdateEvent();
+        event.setCertificateUpdate(certificateUpdate);
+
+        // The subscription request has two types of event responses where the first response sent must be of type
+        // 'SubscribeToCertificateUpdatesResponse' and the following streaming responses must be of type
+        // 'CertificateUpdateEvent'.
+
+        // Since the first streaming response of CertificateUpdateEvent is created before we send the
+        // SubscribeToCertificateUpdatesResponse, we store and send that event in the afterHandleRequest.
+        synchronized (firstStreamingEvent) {
+            if (subscriptionResponseSent.get()) {
+                this.sendStreamEvent(event);
+            } else {
+                firstStreamingEvent.set(event);
+            }
+        }
+    }
+
+    private String getCSRForServer()
+            throws IOException, OperatorCreationException, NoSuchAlgorithmException {
+        this.keyPair = CertificateStore.newRSAKeyPair(RSA_KEY_LENGTH_FOR_SERVER_CERT);
+        return CertificateRequestGenerator
```

Comment:
Let's create a follow up task to clean this up. There's no reason for us to create a CSR here. But that can be done as part of a refactoring task later